File: sdk-node/test/conformance/runner.test.ts
import { readdir, readFile } from 'node:fs/promises'
import { resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

import yaml from 'js-yaml'
import { http, HttpResponse } from 'msw'
import { describe, expect, it } from 'vitest'

import { ThreeCommon } from '@/client'
import * as Errors from '@/errors'

import { setupMockServer, TEST_BASE_URL } from '../helpers/mock-server'

import { dispatchEvents } from './dispatch-events'
import { dispatchInvoices } from './dispatch-invoices'
import { dispatchSubscriptions } from './dispatch-subscriptions'

import type {
  ClientOverrides,
  ExpectedError,
  ExpectedRequest,
  MockResponse,
  Scenario,
  ScenarioCall,
} from './scenario'

const SCENARIOS_DIR = resolve(
  fileURLToPath(import.meta.url),
  '..',
  '..',
  '..',
  '..',
  'conformance',
  'scenarios',
)

async function loadScenarios(): Promise<readonly { file: string; scenario: Scenario }[]> {
  const files: string[] = []
  const walk = async (dir: string): Promise<void> => {
    const entries = await readdir(dir, { withFileTypes: true })
    for (const entry of entries) {
      const full = resolve(dir, entry.name)
      if (entry.isDirectory()) {
        await walk(full)
      } else if (entry.name.endsWith('.yaml')) {
        files.push(full)
      }
    }
  }
  await walk(SCENARIOS_DIR)
  files.sort()

  const scenarios = await Promise.all(
    files.map(async (full) => {
      const text = await readFile(full, 'utf-8')
      // Relative path (e.g. "events/list-happy.yaml") makes it obvious which
      // resource the scenario targets when scrolling test output.
      const file = full.slice(SCENARIOS_DIR.length + 1)
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const scenario: any = yaml.load(text)
      return { file, scenario: scenario as Scenario }
    }),
  )
  return scenarios
}

const server = setupMockServer()

function buildClient(overrides: ClientOverrides = {}): ThreeCommon {
  return new ThreeCommon({
    apiKey: '3co_test',
    baseUrl: TEST_BASE_URL,
    apiVersion: overrides.apiVersion ?? '2026-04-29',
    telemetry: overrides.telemetry ?? true,
    maxRetries: overrides.maxRetries ?? 3,
    retryDelay: { initialMs: 1, maxMs: 4, jitter: false },
  })
}

// dispatchCall routes a scenario call to its per-resource dispatcher.
// Each resource lives in its own file (e.g. dispatch-events.ts,
// dispatch-invoices.ts); add a sibling case here when introducing a new
// resource.
function dispatchCall(
  client: ThreeCommon,
  call: ScenarioCall,
): Promise<unknown> | AsyncIterableIterator<unknown> {
  const resource = call.resource ?? 'events'
  switch (resource) {
    case 'events':
      return dispatchEvents(client, call)
    case 'invoices':
      return dispatchInvoices(client, call)
    case 'subscriptions':
      return dispatchSubscriptions(client, call)
  }
}

function assertRequest(
  observed: Request,
  observedBody: string,
  expected: ExpectedRequest,
  scenarioName: string,
): void {
  expect(observed.method, `[${scenarioName}] method`).toBe(expected.method)

  const url = new URL(observed.url)
  expect(url.pathname, `[${scenarioName}] path`).toBe(expected.path)

  if (expected.query !== undefined) {
    for (const [key, value] of Object.entries(expected.query)) {
      expect(url.searchParams.get(key), `[${scenarioName}] query.${key}`).toBe(value)
    }
  }

  if (expected.headers !== undefined) {
    for (const [key, value] of Object.entries(expected.headers)) {
      expect(observed.headers.get(key), `[${scenarioName}] header.${key}`).toBe(value)
    }
  }

  if (expected.headersAbsent !== undefined) {
    for (const key of expected.headersAbsent) {
      expect(observed.headers.has(key), `[${scenarioName}] header.${key} should be absent`).toBe(
        false,
      )
    }
  }

  if (expected.body !== undefined && expected.body !== null) {
    const parsed = observedBody.length > 0 ? (JSON.parse(observedBody) as unknown) : undefined
    expect(parsed, `[${scenarioName}] body`).toEqual(expected.body)
  }
}

function buildMockResponse(mock: MockResponse): Response {
  return HttpResponse.json(mock.body ?? null, {
    status: mock.status,
    headers: { 'Content-Type': 'application/json', ...(mock.headers ?? {}) },
  })
}

function assertError(err: unknown, expected: ExpectedError, scenarioName: string): void {
  const ExpectedClass = Errors[expected.type]
  expect(err, `[${scenarioName}] error type`).toBeInstanceOf(ExpectedClass)
  const e = err as Errors.ThreeCommonError & {
    retryAfterSeconds?: number
  }
  expect(e.code, `[${scenarioName}] error code`).toBe(expected.code)
  if (expected.httpStatus !== undefined) {
    expect(e.httpStatus, `[${scenarioName}] error httpStatus`).toBe(expected.httpStatus)
  }
  if (expected.requestId !== undefined) {
    expect(e.requestId, `[${scenarioName}] error requestId`).toBe(expected.requestId)
  }
  if (expected.retryAfterSeconds !== undefined) {
    expect(e.retryAfterSeconds, `[${scenarioName}] retryAfterSeconds`).toBe(
      expected.retryAfterSeconds,
    )
  }
  if (expected.details !== undefined) {
    expect(e.details, `[${scenarioName}] error details`).toEqual(expected.details)
  }
}

const allScenarios = await loadScenarios()

describe('cross-language conformance', () => {
  it('loads at least one scenario', () => {
    expect(allScenarios.length).toBeGreaterThan(0)
  })

  it.each(allScenarios)('$scenario.name', async ({ scenario }) => {
    const client = buildClient(scenario.client)
    const observedRequests: { req: Request; body: string }[] = []

    if (scenario.exchanges !== undefined) {
      let exchangeIndex = 0
      const exchanges = scenario.exchanges
      server.use(
        http.all(`${TEST_BASE_URL}${exchanges[0]?.request.path ?? ''}`, async ({ request }) => {
          const body = await request.clone().text()
          observedRequests.push({ req: request, body })
          const exchange = exchanges[exchangeIndex]
          exchangeIndex += 1
          if (exchange === undefined) {
            throw new Error(`unexpected extra request to ${request.url}`)
          }
          return buildMockResponse(exchange.response)
        }),
      )
    } else if (scenario.expectedRequest !== undefined && scenario.mockResponse !== undefined) {
      const expected = scenario.expectedRequest
      const mock = scenario.mockResponse
      server.use(
        http.all(`${TEST_BASE_URL}${expected.path}`, async ({ request }) => {
          const body = await request.clone().text()
          observedRequests.push({ req: request, body })
          return buildMockResponse(mock)
        }),
      )
    } else {
      throw new Error(
        `scenario "${scenario.name}" has no expectedRequest/mockResponse or exchanges`,
      )
    }

    let result: unknown
    let thrown: unknown
    try {
      const dispatched = dispatchCall(client, scenario.call)
      if (scenario.call.method === 'listAutoPaginate') {
        const collected: unknown[] = []
        for await (const item of dispatched as AsyncIterableIterator<unknown>) {
          collected.push(item)
        }
        result = collected
      } else {
        result = await dispatched
      }
    } catch (err) {
      thrown = err
    }

    if (scenario.expectedError !== undefined) {
      expect(thrown, `[${scenario.name}] expected an error`).toBeDefined()
      assertError(thrown, scenario.expectedError, scenario.name)
    } else if (thrown !== undefined) {
      throw thrown instanceof Error
        ? thrown
        : new Error(typeof thrown === 'string' ? thrown : 'unknown thrown value')
    }

    if (scenario.expectedRequest !== undefined && observedRequests[0] !== undefined) {
      assertRequest(
        observedRequests[0].req,
        observedRequests[0].body,
        scenario.expectedRequest,
        scenario.name,
      )
    }

    if (scenario.exchanges !== undefined) {
      scenario.exchanges.forEach((exchange, idx) => {
        const observed = observedRequests[idx]
        if (observed === undefined) {
          throw new Error(`[${scenario.name}] missing observed request at index ${String(idx)}`)
        }
        assertRequest(observed.req, observed.body, exchange.request, scenario.name)
      })
    }

    if (scenario.expectedResult !== undefined) {
      expect(result, `[${scenario.name}] result`).toMatchObject(scenario.expectedResult as object)
    }

    if (scenario.expectedAutoPaginated !== undefined) {
      const items = result as readonly unknown[]
      expect(items, `[${scenario.name}] auto-paginated count`).toHaveLength(
        scenario.expectedAutoPaginated.length,
      )
      scenario.expectedAutoPaginated.forEach((expectedItem, idx) => {
        expect(items[idx], `[${scenario.name}] auto-paginated[${String(idx)}]`).toMatchObject(
          expectedItem,
        )
      })
    }

    if (scenario.expectedCallCount !== undefined) {
      expect(observedRequests.length, `[${scenario.name}] call count`).toBe(
        scenario.expectedCallCount,
      )
    }
  })
})

File: sdk-go/conformance/runner_test.go
// Package conformance runs the shared YAML scenarios at
// ../../conformance/scenarios/**/*.yaml against the Go SDK. Every other SDK in
// this monorepo runs the same scenarios; identical pass/fail across languages
// is the contract. Each resource has its own dispatcher file
// (dispatch_events_test.go, dispatch_invoices_test.go); add a sibling when
// introducing a new resource.
package conformance

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"strings"
	"sync/atomic"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	yaml "gopkg.in/yaml.v3"

	threecommon "github.com/3-Common/sdk/sdk-go"
	"github.com/3-Common/sdk/sdk-go/client"
)

type scenario struct {
	Name string `yaml:"name"`
	Call struct {
		Resource string         `yaml:"resource"`
		Method   string         `yaml:"method"`
		Args     map[string]any `yaml:"args"`
	} `yaml:"call"`
	Client struct {
		MaxRetries *int   `yaml:"maxRetries"`
		APIVersion string `yaml:"apiVersion"`
		Telemetry  *bool  `yaml:"telemetry"`
	} `yaml:"client"`
	ExpectedRequest    *expectedRequest `yaml:"expectedRequest"`
	MockResponse       *mockResponse    `yaml:"mockResponse"`
	Exchanges          []exchange       `yaml:"exchanges"`
	ExpectedResult     map[string]any   `yaml:"expectedResult"`
	ExpectedResultNull bool             `yaml:"expectedResultNull"`
	ExpectedAutoPaged  []map[string]any `yaml:"expectedAutoPaginated"`
	ExpectedError      *expectedError   `yaml:"expectedError"`
	ExpectedCallCount  *int             `yaml:"expectedCallCount"`
}

type expectedRequest struct {
	Method        string            `yaml:"method"`
	Path          string            `yaml:"path"`
	Query         map[string]string `yaml:"query"`
	Headers       map[string]string `yaml:"headers"`
	HeadersAbsent []string          `yaml:"headersAbsent"`
	// Body, when present, is deep-equal compared against the JSON request body
	// (mirroring the Node harness). BodyAbsent asserts no request body was sent.
	Body       any  `yaml:"body"`
	BodyAbsent bool `yaml:"bodyAbsent"`
}

type mockResponse struct {
	Status  int               `yaml:"status"`
	Headers map[string]string `yaml:"headers"`
	Body    any               `yaml:"body"`
}

type exchange struct {
	Request  expectedRequest `yaml:"request"`
	Response mockResponse    `yaml:"response"`
}

type expectedError struct {
	Type              string         `yaml:"type"`
	Code              string         `yaml:"code"`
	HTTPStatus        int            `yaml:"httpStatus"`
	RequestID         string         `yaml:"requestId"`
	RetryAfterSeconds *float64       `yaml:"retryAfterSeconds"`
	Details           map[string]any `yaml:"details"`
}

func TestConformance(t *testing.T) {
	t.Parallel()

	scenarios, err := loadScenarios("../../conformance/scenarios")
	require.NoError(t, err)
	require.NotEmpty(t, scenarios, "no conformance scenarios found")

	for _, sc := range scenarios {
		// sc.path is "events/list-happy.yaml" — keeps the resource visible
		// when reading test output.
		t.Run(sc.path, func(t *testing.T) {
			t.Parallel()
			runScenario(t, sc.scn)
		})
	}
}

type loadedScenario struct {
	path string
	scn  scenario
}

func loadScenarios(dir string) ([]loadedScenario, error) {
	var out []loadedScenario
	walkErr := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() || !strings.HasSuffix(d.Name(), ".yaml") {
			return nil
		}
		bytes, err := os.ReadFile(path)
		if err != nil {
			return fmt.Errorf("read %s: %w", path, err)
		}
		var sc scenario
		if err := yaml.Unmarshal(bytes, &sc); err != nil {
			return fmt.Errorf("parse %s: %w", path, err)
		}
		// Relative path (e.g. "events/list-happy.yaml") keeps the resource
		// visible in the t.Run subtest name.
		rel, relErr := filepath.Rel(dir, path)
		if relErr != nil {
			rel = path
		}
		out = append(out, loadedScenario{path: rel, scn: sc})
		return nil
	})
	if walkErr != nil {
		return nil, walkErr
	}
	return out, nil
}

func runScenario(t *testing.T, sc scenario) {
	t.Helper()

	exchanges := sc.Exchanges
	if len(exchanges) == 0 && sc.MockResponse != nil {
		req := expectedRequest{}
		if sc.ExpectedRequest != nil {
			req = *sc.ExpectedRequest
		}
		exchanges = []exchange{{Request: req, Response: *sc.MockResponse}}
	}

	var calls atomic.Int32
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		idx := int(calls.Add(1)) - 1
		require.Less(t, idx, len(exchanges), "%s: more requests than expected", sc.Name)

		ex := exchanges[idx]
		assertRequestMatches(t, sc.Name, ex.Request, r)
		writeMockResponse(w, ex.Response)
	}))
	defer srv.Close()

	cfg := threecommon.Config{
		APIKey:     "3co_test",
		BaseURL:    srv.URL,
		HTTPClient: srv.Client(),
	}
	if sc.Client.MaxRetries != nil {
		cfg.MaxRetries = sc.Client.MaxRetries
	}
	if sc.Client.APIVersion != "" {
		cfg.APIVersion = sc.Client.APIVersion
	}
	if sc.Client.Telemetry != nil {
		cfg.Telemetry = sc.Client.Telemetry
	}

	api, err := client.New(cfg)
	require.NoError(t, err)

	result, callErr := dispatch(t, api, sc)

	// Validate result / error.
	switch {
	case sc.ExpectedError != nil:
		require.Error(t, callErr, "%s: expected error", sc.Name)
		assertExpectedError(t, sc.Name, *sc.ExpectedError, callErr)
	case sc.ExpectedAutoPaged != nil:
		require.NoError(t, callErr, "%s: %v", sc.Name, callErr)
		assertAutoPaginated(t, sc.Name, sc.ExpectedAutoPaged, result)
	case sc.ExpectedResult != nil:
		require.NoError(t, callErr, "%s: %v", sc.Name, callErr)
		assertJSONShape(t, sc.Name, sc.ExpectedResult, result)
	case sc.ExpectedResultNull:
		require.NoError(t, callErr, "%s: %v", sc.Name, callErr)
		assert.Nil(t, result, "%s: result should be nil", sc.Name)
	}

	if sc.ExpectedCallCount != nil {
		assert.Equal(t, int32(*sc.ExpectedCallCount), calls.Load(), "%s: call count", sc.Name)
	}
}

// dispatch routes a scenario call to the appropriate resource dispatcher.
// Each resource lives in its own file (e.g. dispatch_events_test.go,
// dispatch_invoices_test.go); add a sibling case here when introducing a new
// resource.
func dispatch(t *testing.T, api *client.API, sc scenario) (any, error) {
	t.Helper()
	ctx := context.Background()

	resource := sc.Call.Resource
	if resource == "" {
		resource = "events"
	}

	switch resource {
	case "events":
		return dispatchEvents(t, api, ctx, sc)
	case "invoices":
		return dispatchInvoices(t, api, ctx, sc)
	case "subscriptions":
		return dispatchSubscriptions(t, api, ctx, sc)
	case "contacts":
		return dispatchContacts(t, api, ctx, sc)
	case "entitlements":
		return dispatchEntitlements(t, api, ctx, sc)
	case "prices":
		return dispatchPrices(t, api, ctx, sc)
	case "features":
		return dispatchFeatures(t, api, ctx, sc)
	case "forms":
		return dispatchForms(t, api, ctx, sc)
	case "properties":
		return dispatchProperties(t, api, ctx, sc)
	}
	t.Fatalf("unsupported scenario resource %q", resource)
	return nil, nil
}

func anyToIntPtr(v any) *int {
	switch n := v.(type) {
	case int:
		return &n
	case int64:
		i := int(n)
		return &i
	case float64:
		i := int(n)
		return &i
	}
	return nil
}

func assertRequestMatches(t *testing.T, scenarioName string, want expectedRequest, r *http.Request) {
	t.Helper()
	if want.Method != "" {
		assert.Equal(t, want.Method, r.Method, "%s: method", scenarioName)
	}
	if want.Path != "" {
		assert.Equal(t, want.Path, r.URL.Path, "%s: path", scenarioName)
	}
	for k, v := range want.Query {
		assert.Equal(t, v, r.URL.Query().Get(k), "%s: query[%s]", scenarioName, k)
	}
	for k, v := range want.Headers {
		assert.Equal(t, v, r.Header.Get(k), "%s: header[%s]", scenarioName, k)
	}
	for _, k := range want.HeadersAbsent {
		assert.Empty(t, r.Header.Get(k), "%s: header %s should be absent", scenarioName, k)
	}

	// Body assertion mirrors the Node harness: a non-nil want.Body is
	// deep-equal compared against the parsed JSON request body, so an SDK that
	// drops a field, sends an extra one, or omits an explicit null fails here.
	raw, _ := io.ReadAll(r.Body)
	if want.Body != nil {
		var got any
		if len(raw) > 0 {
			require.NoError(t, json.Unmarshal(raw, &got), "%s: parse request body", scenarioName)
		}
		assert.Equal(t, normalize(want.Body), normalize(got), "%s: body", scenarioName)
	}
	if want.BodyAbsent {
		assert.Empty(t, raw, "%s: request body should be absent", scenarioName)
	}
}

func writeMockResponse(w http.ResponseWriter, m mockResponse) {
	for k, v := range m.Headers {
		w.Header().Set(k, v)
	}
	status := m.Status
	if status == 0 {
		status = http.StatusOK
	}
	w.WriteHeader(status)
	if m.Body != nil {
		_ = json.NewEncoder(w).Encode(m.Body)
	}
}

func assertExpectedError(t *testing.T, scenarioName string, want expectedError, got error) {
	t.Helper()

	switch want.Type {
	case "ThreeCommonAuthError":
		var e *threecommon.AuthError
		require.True(t, errors.As(got, &e), "%s: want AuthError, got %T", scenarioName, got)
		assertAPIErrorFields(t, scenarioName, want, e.APIError)
	case "ThreeCommonPermissionError":
		var e *threecommon.PermissionError
		require.True(t, errors.As(got, &e), "%s: want PermissionError, got %T", scenarioName, got)
		assertAPIErrorFields(t, scenarioName, want, e.APIError)
	case "ThreeCommonNotFoundError":
		var e *threecommon.NotFoundError
		require.True(t, errors.As(got, &e), "%s: want NotFoundError, got %T", scenarioName, got)
		assertAPIErrorFields(t, scenarioName, want, e.APIError)
	case "ThreeCommonValidationError":
		var e *threecommon.ValidationError
		require.True(t, errors.As(got, &e), "%s: want ValidationError, got %T", scenarioName, got)
		assertAPIErrorFields(t, scenarioName, want, e.APIError)
	case "ThreeCommonConflictError":
		var e *threecommon.ConflictError
		require.True(t, errors.As(got, &e), "%s: want ConflictError, got %T", scenarioName, got)
		assertAPIErrorFields(t, scenarioName, want, e.APIError)
	case "ThreeCommonRateLimitError":
		var e *threecommon.RateLimitError
		require.True(t, errors.As(got, &e), "%s: want RateLimitError, got %T", scenarioName, got)
		assertAPIErrorFields(t, scenarioName, want, e.APIError)
		if want.RetryAfterSeconds != nil {
			assert.InDelta(t, *want.RetryAfterSeconds, e.RetryAfter.Seconds(), 0.001, "%s: retryAfter", scenarioName)
		}
	case "ThreeCommonServerError":
		var e *threecommon.ServerError
		require.True(t, errors.As(got, &e), "%s: want ServerError, got %T", scenarioName, got)
		assertAPIErrorFields(t, scenarioName, want, e.APIError)
	default:
		t.Fatalf("%s: unsupported expectedError.type %q", scenarioName, want.Type)
	}
}

func assertAPIErrorFields(t *testing.T, scenarioName string, want expectedError, got *threecommon.APIError) {
	t.Helper()
	if want.Code != "" {
		assert.Equal(t, want.Code, got.Code, "%s: error.code", scenarioName)
	}
	if want.HTTPStatus != 0 {
		assert.Equal(t, want.HTTPStatus, got.HTTPStatus, "%s: error.httpStatus", scenarioName)
	}
	if want.RequestID != "" {
		assert.Equal(t, want.RequestID, got.RequestID, "%s: error.requestId", scenarioName)
	}
	for k, v := range want.Details {
		assert.Equal(t, normalize(v), normalize(got.Details[k]), "%s: error.details[%s]", scenarioName, k)
	}
}

func assertJSONShape(t *testing.T, scenarioName string, want map[string]any, got any) {
	t.Helper()
	gotJSON, err := json.Marshal(got)
	require.NoError(t, err)

	var gotMap map[string]any
	require.NoError(t, json.Unmarshal(gotJSON, &gotMap))

	for k, wantVal := range want {
		assertSubset(t, fmt.Sprintf("%s: result[%s]", scenarioName, k), wantVal, gotMap[k])
	}
}

func assertAutoPaginated(t *testing.T, scenarioName string, want []map[string]any, got any) {
	t.Helper()
	gotJSON, err := json.Marshal(got)
	require.NoError(t, err)

	var gotList []map[string]any
	require.NoError(t, json.Unmarshal(gotJSON, &gotList))

	require.Equal(t, len(want), len(gotList), "%s: auto-paginated length", scenarioName)
	for i := range want {
		for k, wantVal := range want[i] {
			assertSubset(t, fmt.Sprintf("%s: paged[%d][%s]", scenarioName, i, k), wantVal, gotList[i][k])
		}
	}
}

// assertSubset checks whether want is contained in got.
//
// For maps, every key in want must also exist in got with the same value.
// For slices, each element in want is compared with the element at the same
// index in got.
// Scalar values are compared directly after numeric normalization.
//
// got can contain extra fields or elements that aren’t present in want.
func assertSubset(t *testing.T, prefix string, want, got any) {
	t.Helper()

	switch w := normalize(want).(type) {
	case map[string]any:
		gm, ok := normalize(got).(map[string]any)
		if !ok {
			t.Errorf("%s: want object, got %T", prefix, got)
			return
		}
		for k, v := range w {
			assertSubset(t, prefix+"."+k, v, gm[k])
		}
	case []any:
		gs, ok := normalize(got).([]any)
		if !ok {
			t.Errorf("%s: want array, got %T", prefix, got)
			return
		}
		require.Equal(t, len(w), len(gs), "%s: array length", prefix)
		for i, v := range w {
			assertSubset(t, fmt.Sprintf("%s[%d]", prefix, i), v, gs[i])
		}
	default:
		assert.Equal(t, w, normalize(got), prefix)
	}
}

// normalize converts scalars and nested values into a form we can compare.
//
// YAML decoding can produce int, float64, string, []any, and map[string]any.
// JSON decoding produces the same types, except numbers come through as
// float64.
//
// To keep comparisons consistent, ints are converted to float64.
func normalize(v any) any {
	switch x := v.(type) {
	case int:
		return float64(x)
	case int32:
		return float64(x)
	case int64:
		return float64(x)
	case []any:
		out := make([]any, len(x))
		for i, item := range x {
			out[i] = normalize(item)
		}
		return out
	case map[string]any:
		out := make(map[string]any, len(x))
		for k, item := range x {
			out[k] = normalize(item)
		}
		return out
	default:
		return v
	}
}
